File: python-wrapper/examples/mcp_agent.py
import asyncio
import os
from ferrant import Agent, McpTools

async def main() -> None:
    mcp = await McpTools.connect("npx", ["-y", "@modelcontextprotocol/server-filesystem", "."])
    print("Discovered:", mcp.names())
    agent = Agent.openai("gpt-5-nano", os.environ["OPENAI_API_KEY"],
        instructions="Inspect files but never modify them unless asked.", mcp=mcp)
    print(await agent.run("Summarize this project's Rust source files."))

asyncio.run(main())

File: examples/mcp_agent.rs
//! Agent using tools discovered from an MCP filesystem server.
//! Requires Node.js/npx and an OPENAI_API_KEY.

use ferrant::llm::openai::OpenAiModel;
use ferrant::{Agent, McpClient};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    dotenvy::dotenv().ok();
    let server = McpClient::connect(
        "npx",
        ["-y", "@modelcontextprotocol/server-filesystem", "."],
    )
    .await?;

    let model = OpenAiModel::new("gpt-5-nano", std::env::var("OPENAI_API_KEY")?);
    let mut agent = Agent::builder(model)
        .instructions("Use the MCP filesystem tools to inspect the project. Never modify files unless explicitly asked.")
        .tools(server.tools().await?)
        .build();

    println!(
        "{}",
        agent
            .run("Summarize the Rust source files in this project.")
            .await?
    );
    Ok(())
}
